src/structs/multiplexer.rs

```diff
@@ -270,151 +293,200 @@ impl MultiplexReader {
+            let (tx, rx) = channel::<Message>();
             channel_to_sender.push(tx);
 
-            my_spawn(s, move || {
+            let cloned_processor = arc_processor.clone();
+            let cloned_result_sender = result_sender.clone();
+
+            my_spawn_simple(move || {
                 // get the appropriate receiver so we can read out data from it
                 let mut proc_reader = MultiplexReader {
                     thread_id: thread_id as u8,
                     current_buffer: Cursor::new(Vec::new()),
                     receiver: rx,
                     end_of_file: false,
                 };
-                *result = Some(processor(thread_id, &mut proc_reader));
+
+                // nothing to do if we fail to return the results (since the caller
+                // died and we can't return the results)
+                _ = cloned_result_sender
+                    .send((thread_id, cloned_processor(thread_id, &mut proc_reader)));
             });
         }
 
-        // now that the channels are waiting for input, read the stream and send all the buffers to their respective readers
-        loop {
-            let mut thread_marker_a = [0; 1];
-            if reader.read(&mut thread_marker_a)? == 0 {
-                break;
-            }
-
-            let thread_marker = thread_marker_a[0];
-
-            let thread_id = (thread_marker & 0xf) as u8;
-
-            if thread_id >= channel_to_sender.len() as u8 {
-                return err_exit_code(
-                    ExitCode::BadLeptonFile,
-                    format!("invalid thread_id {0}", thread_id).as_str(),
-                );
-            }
-
-            let data_length = if thread_marker < 16 {
-                let b0 = reader.read_u8().context(here!())?;
-                let b1 = reader.read_u8().context(here!())?;
-
-                ((b1 as usize) << 8) + b0 as usize + 1
-            } else {
-                // This format is used by Lepton C++ to write encoded chunks with length of 4096, 16384 or 65536 bytes
-                let flags = (thread_marker >> 4) & 3;
-
-                1024 << (2 * flags)
-            };
-
-            //info!("offset {0} len {1}", reader.stream_position()?-2, data_length);
-
-            let mut buffer = vec![0; data_length as usize];
-
-            reader
-                .read_exact(&mut buffer)
-                .with_context(|| format!("reading {0} bytes", buffer.len()))?;
-
-            let e =
-                channel_to_sender[thread_id as usize].send(Message::WriteBlock(thread_id, buffer));
+        MultiplexReaderState {
+            sender_channels: channel_to_sender,
+            result_receiver: result_receiver,
+            current_state: State::StartBlock,
+            reserve,
+        }
+    }
 
-            if let Err(e) = e {
-                error_sending = Some(e);
+    pub fn process_to_end(&mut self, source: &mut impl BufRead) -> Result<Vec<u8>> {
```

Comment:
Is it used anywhere?

Reply:
Used by the utility for dumping the code.. moved it there